### PR TITLE
Add the memory estimator tool

### PR DESCRIPTION
**Pull request by aggarg**
Description
-----------
This tool helps in determining the memory estimates for FreeRTOS libraries.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aggarg** [comment]
> 
> 
> We should make this as compatible with Python 3 as possible, since 2 is EOL

Agree. Made the changes and tested with both python 2 and 3. Thanks for the suggestion.

File: tools/memory_estimator/memory_estimator.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    print('---- Starting Build. Compiler: {} ---- \n'.format(compiler))
+
+    proc = subprocess.Popen(["make", compiler_command], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+    warnings = []
+    for line in iter(proc.stdout.readline, b''):
+        line = line.strip()
+        print(line)
+        if re.match('.*: warning: .*', line):
```
[the remark this answers is not in the document]
Comment (aggarg) [reply tools/memory_estimator/memory_estimator.py:110]: Done.

File: tools/memory_estimator/memory_estimator.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        for warning in warnings:
+            print(warning)
+        print('------------------')
+
+
+def pretty_print_sizes(sizes):
+    print('\n\n---- Sizes ---- ')
+    print('---------------')
+    for size in sizes[1:]:
+        print(size)
```
[the remark this answers is not in the document]
Comment (aggarg) [reply tools/memory_estimator/memory_estimator.py:157]: Though it was not generating any error with python 2 or 3 but I agree that it is better to decode as 'utf-8' first. Done.

File: tools/memory_estimator/memory_estimator.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    return vars(parser.parse_args())
+
+
+
+def main():
+    args = parse_arguments()
+
+    # Generate Makefile.
+    generate_makefile(args['lts_path'], args['optimization'], args['lib'])
+
```
[the remark this answers is not in the document]
Comment (aggarg) [reply tools/memory_estimator/memory_estimator.py:180]: Added `make -f __MAKE_FILE_OUTPUT__` which achieves the same effect of making the script runnable from any directory without explicitly changing cwd.

File: tools/memory_estimator/memory_estimator.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
+import io
```
[the remark this answers is not in the document]
Comment (aggarg) [reply tools/memory_estimator/memory_estimator.py:3]: Done.

File: tools/memory_estimator/memory_estimator.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    generate_makefile(args['lts_path'], args['optimization'], args['lib'])
+
+    # Run make build.
+    warnings = make(args['compiler'])
+
+    # Run make size.
+    calculated_sizes = calculate_size(args['sizetool'])
+
+    # Remove the generated artifacts.
+    if args['dontclean'] == False:
```
[the remark this answers is not in the document]
Comment (aggarg) [reply tools/memory_estimator/memory_estimator.py]: Done.

File: tools/memory_estimator/memory_estimator.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        print(line)
+
+
+def remove_generated_makefile():
+    os.remove(__GENERATED_MAKE_FILE__)
+
+
+def pretty_print_warnings(warnings):
+    if len(warnings) > 0:
+        print('\n\n---- Warnings ---- ')
```
Comment (yuhui-zheng) [inline tools/memory_estimator/memory_estimator.py:146]: I would recommend logging module in the next round. https://docs.python.org/3/library/logging.html. 

The benefit is you could customize log line formatting in a cleaning way, and also having more than one streams (e.g. log to both screen and file). And the file can go directly to infra for further parsing or entering into database...

**Pull request merged**